### PR TITLE
GAPI Fluid: SIMD for ConvertTo.

### PR DESCRIPTION
**Pull request by anna-khakimova**
SIMD for ConvertTo

Performance:
[FluidConvertToSIMD.xlsx](https://github.com/opencv/opencv/files/8349633/FluidConvertToSIMD.xlsx)

<cut/>

```
force_builders=Linux AVX2,Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
Xbuild_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

build_image:Custom=centos:7
buildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

Xbuild_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.2.0

buildworker:Custom Win=windows-3

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
Xtest_opencl:Custom=OFF
Xtest_bigdata:Custom=1
Xtest_filter:Custom=*

CPU_BASELINE:Custom Win=AVX512_SKX
CPU_BASELINE:Custom=SSE4_2
```

### CODE REVIEW

File: modules/gapi/src/backends/fluid/gfluidcore_func.dispatch.cpp [hunk 1]
```diff
@@ -344,6 +342,59 @@ SUB_SIMD(float, float)
...
 #undef SUB_SIMD
 
+#define CONVERTTO_NOCOEF_SIMD(SRC, DST)                              \
+int convertto_simd(const SRC in[], DST out[], const int length)      \
+{                                                                    \
+    CV_CPU_DISPATCH(convertto_simd, (in, out, length),               \
+                    CV_CPU_DISPATCH_MODES_ALL);                      \
+}
+
+//CONVERTTO_NOCOEF_SIMD(uchar, uchar)
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidcore_func.dispatch.cpp:352]: why is it commented?

File: modules/gapi/src/backends/fluid/gfluidcore_func.simd.hpp [hunk 2]
```diff
@@ -2778,6 +2823,327 @@ SUB_SIMD(float, float)
...
+    v_int16 r1 = v_pack(a1, a2);
+    v_int16 r2 = v_pack(a3, a4);
+
+    vx_store(outx, v_pack_u(r1, r2));
+}
+
+template<typename SRC>
+CV_ALWAYS_INLINE
+typename std::enable_if<std::is_same<SRC, short>::value ||
+                        std::is_same<SRC, ushort>::value, void>::type
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidcore_func.simd.hpp:2870]: it takes a lot of time on review to expand this conditions in mind. Also i see that  these SFINAE repeat again and again

I suggest to introduce MACRO with such condition incapsulation in the way, eventually:
```
typename std::enable_if<SHORT_OR_USHORT>::type 
convertto_simd_nocoeff_impl(...)
<and other functions with similar SFINAE> 
...
#define SHORT_OR_USHORT std::is_same<SRC, short>::value ||  std::is_same<SRC, ushort>::value
```
etc

Comment (anna-khakimova) [reply modules/gapi/src/backends/fluid/gfluidcore_func.simd.hpp]: Applied.

**Review by sivanov-work — APPROVED** [review]
LGTM
Can we apply MACRO in SFINAE approach  for other operation dispatchering functions  in next  PRs?

**Comment by anna-khakimova** [comment]
> LGTM Can we apply MACRO in SFINAE approach for other operation dispatchering functions in next PRs?

There is sense to do it in the separate PR in the bounds of a refactoring task.

**Review by dmatveev — APPROVED** [review]
👍

**Comment by anna-khakimova** [comment]
@alalek Could you please take a look this PR and merge if it looks good for you?

**Pull request merged**